src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs

```diff
@@ -459,8 +481,459 @@ public void WithAnalyzerConfigDocumentTextLoader()
             Assert.Throws<ArgumentNullException>(() => solution.WithAnalyzerConfigDocumentTextLoader(null!, loader, PreservationMode.PreserveIdentity));
             Assert.Throws<InvalidOperationException>(() => solution.WithAnalyzerConfigDocumentTextLoader(s_unrelatedDocumentId, loader, PreservationMode.PreserveIdentity));
         }
+
+        [Fact]
+        public void WithProjectAssemblyName()
+        {
+            var projectId = ProjectId.CreateNewId();
+
+            var solution = CreateSolution()
+                .AddProject(projectId, "proj1", "proj1.dll", LanguageNames.CSharp);
+
+            // any character is allowed
+            var assemblyName = "\0<>a/b/*.dll";
+
+            var newSolution = solution.WithProjectAssemblyName(projectId, assemblyName);
+            Assert.Equal(assemblyName, newSolution.GetProject(projectId)!.AssemblyName);
+
+            Assert.Same(newSolution, newSolution.WithProjectAssemblyName(projectId, assemblyName));
+
+            Assert.Throws<ArgumentNullException>("assemblyName", () => solution.WithProjectAssemblyName(projectId, null!));
+            Assert.Throws<ArgumentNullException>("projectId", () => solution.WithProjectAssemblyName(null!, "x.dll"));
+            Assert.Throws<InvalidOperationException>(() => solution.WithProjectAssemblyName(ProjectId.CreateNewId(), "x.dll"));
```

Comment:
Not an ArgumentException, since the argument is wrong? (it's not anything to do with "the object's current state")

Reply:
That's not how it works today. All the checks that project id is part of the solution throw InvalidOperationException. I don't particularly like it but changing that would be breaking.